packages/devtools_app/lib/src/screens/performance/rebuild_counts.dart

```diff
@@ -20,112 +21,297 @@ import '../../primitives/utils.dart';
+
+  /// Source of truth for all resolution fo location ids to [Location] objects.
+  final LocationMap locationMap = LocationMap();
+
+  /// Map from frame id to list of rebuilds for the frame.
+  final _rebuildsForFrame = <int, List<RebuildLocation>>{};
+
+  final _stats = _RebuildCountStats();
+  ValueListenable<List<RebuildLocation>> get locationStats => _stats.locations;
+
+  List<RebuildLocation>? get rebuildsForLastFrame =>
+      _rebuildsForFrame.values.lastOrNull;
+
+  List<RebuildLocation>? rebuildsForFrame(int frameNumber) {
+    return _rebuildsForFrame[frameNumber];
+  }
+
+  Map<String, dynamic>? toJson() {
+    if (_rebuildsForFrame.isEmpty) {
+      // No need to encode data unless there were actually rebuilds reported.
+      return null;
+    }
+    final frames = <dynamic>[];
+
+    _rebuildsForFrame.forEach((id, rebuilds) {
+      final events = <int>[];
+      for (RebuildLocation rebuild in rebuilds) {
+        events
+          ..add(rebuild.location.id)
+          ..add(rebuild.buildCount);
       }
+      frames.add({'frameNumber': id, 'events': events});
+    });
+    return <String, dynamic>{
+      'locations': locationMap.toJson(),
+      'frames': frames,
+    };
+  }
+
+  void processRebuildEvent(Map<String, dynamic> json) {
+    // parse locations
+    if (json.containsKey('locations')) {
+      locationMap.processLocationMap(json['locations'] as Map<String, dynamic>);
     }
 
+    processRebuildsForFrame(json);
+
+    // We've updated the build counts and possibly the locations.
+    _stats.dataChanged();
+  }
+
+  void clearFromRestart() {
+    _rebuildsForFrame.clear();
+    locationMap.clear();
+    _stats.clear();
+  }
+
+  void clearAllCounts() {
+    _stats.clear();
+    _rebuildsForFrame.clear();
+  }
+
+  void processRebuildsForFrame(Map<String, dynamic> json) {
+    final int frameNumber = json['frameNumber'];
     // parse events
     final List<int> events = (json['events'] as List).cast<int>();
+    final rebuildsForFrame = <RebuildLocation>[];
     for (int i = 0; i < events.length; i += 2) {
       final id = events[i];
       final count = events[i + 1];
 
-      final location = _locationMap[id];
-      location?.buildCount += count;
+      final location = locationMap[id];
+      rebuildsForFrame.add(RebuildLocation(location, buildCount: count));
+      _stats.incrementCount(location, count);
+    }
+    _rebuildsForFrame[frameNumber] = rebuildsForFrame;
+    while (_rebuildsForFrame.length > rebuildFrameCacheSize) {
+      _rebuildsForFrame.remove(_rebuildsForFrame.keys.first);
     }
-
-    // We've updated the build counts and possibly the locations.
-    _locations.notifyListeners();
   }
+}
 
-  void clearFromReload() {
-    // TODO(devoncarew): We need to call this when we see a hot reload or restart.
+class Location {
+  Location({required this.id, this.path, this.line, this.column, this.name});
 
-    _locationMap.clear();
-    _locations.clear();
-  }
+  int id;
```

Comment:
should this be final?

Reply:
done here and elsewhere. A few things were final but I'd missed marking them. @srawlins this is a case where I'd love whole package level lints to warn me that I had fields that could be final but I'd forgotten to mark final.